toolkit/offchain/src/register.rs

```diff
@@ -1,14 +1,168 @@
+	A: AwaitTx,
+>(
+	genesis_utxo: UtxoId,
+	block_producer_registration: &BlockProducerRegistration,
+	payment_signing_key: MainchainPrivateKey,
+	ogmios_client: &C,
+	await_tx: A,
+) -> anyhow::Result<Option<OgmiosTx>> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key.0, ogmios_client).await?;
+	let validator = crate::scripts_data::registered_candidates_scripts(genesis_utxo)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let registration_utxo = ctx
+		.payment_key_utxos
+		.iter()
+		.find(|u| u.to_domain() == block_producer_registration.registration_utxo)
+		.ok_or(anyhow!("registration utxo not found at payment address"))?;
+	let all_registration_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
+	let own_registrations = get_own_registrations(
+		block_producer_registration.own_pkh,
+		block_producer_registration.stake_ownership.pub_key.clone(),
+		&all_registration_utxos,
+	)?;
+	let own_registration_utxos = own_registrations.iter().map(|r| r.0.clone()).collect::<Vec<_>>();
+
+	if own_registrations
+		.iter()
+		.any(|(_, existing_registration)| block_producer_registration == existing_registration)
+	{
+		log::info!("✅ Block producer already registered with same keys.");
+		return Ok(None);
+	}
+
+	let zero_ex_units = ExUnits::new(&0u64.into(), &0u64.into());
+	let tx = register_tx(
+		&validator,
+		block_producer_registration,
+		registration_utxo,
+		&own_registration_utxos,
+		&ctx,
+		zero_ex_units,
+	)?;
+
+	let evaluate_response =
+		ogmios_client.evaluate_transaction(&tx.to_bytes()).await.map_err(|e| {
+			anyhow!(
+				"Evaluate candidate registration transaction request failed: {}, bytes: {}",
+				e,
+				hex::encode(tx.to_bytes())
+			)
+		})?;
+	let validator_redeemer_ex_units = get_first_validator_budget(evaluate_response)?;
+	let tx = register_tx(
+		&validator,
+		block_producer_registration,
+		registration_utxo,
+		&own_registration_utxos,
+		&ctx,
+		validator_redeemer_ex_units,
+	)?;
+	let signed_tx = ctx.sign(&tx).to_bytes();
+	let result = ogmios_client.submit_transaction(&signed_tx).await.map_err(|e| {
+		anyhow!(
+			"Submit candidate registration transaction request failed: {}, bytes: {}",
+			e,
+			hex::encode(tx.to_bytes())
+		)
+	})?;
+	let tx_id = result.transaction.id;
+	log::info!("✅ Transaction submited. ID: {}", hex::encode(result.transaction.id));
+	await_tx
+		.await_tx_output(ogmios_client, UtxoId { tx_hash: McTxHash(tx_id), index: UtxoIndex(0) })
+		.await?;
+
+	Ok(Some(result.transaction))
+}
+
+fn get_own_registrations(
+	own_pkh: MainchainAddressHash,
+	spo_pub_key: MainchainPublicKey,
+	validator_utxos: &[OgmiosUtxo],
+) -> Result<Vec<(OgmiosUtxo, BlockProducerRegistration)>, anyhow::Error> {
+	let mut own_registrations = Vec::new();
+	for validator_utxo in validator_utxos {
+		let datum = validator_utxo.datum.clone().ok_or_else(|| {
+			anyhow!("Invalid state: an UTXO at the validator script address does not have a datum")
```

Comment:
Is the presence of datum checked by smart contract? Is its shape checked as well? I think you should instead ignore such UTXOs but for sure don't crash this method because some malicious actor can transfer UTXO that will prevent anyone from registering with this code.